scripts/CatapultRecipeUpdater.py

```diff
@@ -0,0 +1,191 @@
+def dispatch_subprocess(command_line, cwd=None, handle_error=True):
+	print(' '.join(command_line))
+	result = subprocess.run(command_line, check=False, cwd=cwd, capture_output=True)
+	if handle_error and 0 != result.returncode:
+		raise subprocess.SubprocessError(f'command failed with exit code {result.returncode}\n{result}')
+
+	output = result.stdout.decode('utf-8') if 0 == result.returncode else result.stderr.decode('utf-8')
+	if output:
+		print(output)
+
+	return output, result.returncode
+
+
+class CatapultRecipesUpdater:
+	def __init__(self, source_path):
+		self.source_path = Path(source_path).absolute()
+
+	@staticmethod
+	async def _initialize_conan():
+		dispatch_subprocess(['conan', 'config', 'init'])
+		dispatch_subprocess(['conan', 'config', 'set', 'general.revisions_enabled=1'])
+		dispatch_subprocess(['conan', 'profile', 'update', 'settings.compiler.libcxx=libstdc++11', 'default'])
+		dispatch_subprocess(['conan', 'remote', 'add', '--force', 'nemtech', CONAN_NEMTECH_REMOTE])
+
+	@staticmethod
+	async def _get_recipe_latest_version(owner, repo):
+		url = f'https://api.github.com/repos/{owner}/{repo}/releases/latest'
+		async with ClientSession(raise_for_status=True) as session:
+			async with session.get(url) as response:
+				response_json = await response.json()
+				if 200 != response.status:
+					raise Exception(f'failed to get latest release for {owner}/{repo} {response_json}')
+				return re.sub(r'[^\d\.]', '', response_json['tag_name'])
+
+	@staticmethod
+	async def _read_yaml_file(filepath):
+		with open(filepath, 'r') as file:
+			return yaml.safe_load(file)
+
+	async def _get_current_version(self, name):
+		recipe_config_filepath = self.source_path / f'{name}/config.yml'
+		config = await self._read_yaml_file(recipe_config_filepath)
+		return next(iter(config['versions'].keys()))
+
+	async def _get_update_if_available(self, recipe):
+		owner = RECIPES_REPOSITORIES[recipe]
+		recipe_update = {}
+		latest_version = await self._get_recipe_latest_version(owner, recipe)
+		recipe = 'zeromq' if 'libzmq' == recipe else recipe
+		current_version = await self._get_current_version(recipe)
+		print(f'checking recipe {recipe} {current_version} -> {latest_version}')
+		if semver.compare(latest_version, current_version) > 0:
+			print(f'{recipe} has new version: {latest_version}')
+			recipe_update[recipe] = (current_version, latest_version)
+
+		return recipe_update
+
+	async def get_available_update(self, recipes):
+		results = await asyncio.gather(*[self._get_update_if_available(recipe) for recipe in recipes])
+		recipe_with_updates = dict(ChainMap(*results))
+		return recipe_with_updates
+
+	async def _update_conandata_sha(self, recipe, filepath, version):
+		config = await self._read_yaml_file(filepath)
+		url = config['sources'][version]['url']
+		with tempfile.TemporaryDirectory() as tmpdir:
+			download_file = f'{tmpdir}/{recipe}.{version}.tar.gz'
+			dispatch_subprocess(['curl', '-LJ', url, '--output', f'{download_file}'])
```

Comment:
just `download_file` ?

Reply:
do you need `f'{tmp_source_filepath}` or would just `tmp_source_filepath` work below (2x)?